### PR TITLE
Custom masking function

### PR DESCRIPTION
**Pull request by oliverdutton**
I have cases with different types of atoms which don't interact. I would like to minimise the size of the pairlist buffer at setup so have implemented an optional function.

### CODE REVIEW

**Comment by oliverdutton** [comment]
I have now signed CLA, please rescan

**Comment by oliverdutton** [comment]
Having used it I now seen the sparse neighbor list case too.

Index of particle 1 is given by the index of the array in the first dimension.

Index of particle 2 is given by the value in the array.

I will update the documentation slightly as I had said ‘square’ array

**Comment by sschoenholz** [comment]
Thanks for the contribution, I think a custom mask function could make a lot of sense. 

A few quick comments:
1) Neighbor lists are constructed using a few phases. First, a "broad phase" in which candidate neighbors are identified (possibly using cell-lists) and then a pruning step in which only candidates that are within the specified cutoff are included. Currently, the mask is applied in between the broad phase and the pruning phase. However, the pruning phase is also the step that converts the neighbor list into the appropriate format. Therefore, as written the mask is always applied to neighbor lists whose format resembles the dense case. (It is useful to apply the mask before pruning because then the capacity is properly estimated). Therefore, I actually think that your original documentation was correct!

2) Can you include a new test following the example either in energy_test.py or partition_test.py that exercises the custom mask functionality?

Once these changes are addressed I'm more than happy to merge. Thanks again!

**Comment by oliverdutton** [comment]
Thank you for the speedy reply. 

Replying to your comments:
  - When cell list is enabled and a sparse representation is used, idx is not square.
  When you have cell list disabled, it's square.
  In the image below, the top two are from cell list on, bottom one is cell list disabled.
- Will add tests

  This is from the notebook in #168 and may link to it.
  image

**Comment by sschoenholz** [comment]
Thanks! Just to clarify, whether or not `idx` is square has to do with the use of cell-lists rather than the format of the neighbor list. When cell lists aren't used `idx` is an (N,N) matrix containing every pair of ids. This list is then pruned, but it costs O(N^2) time. When using cell-lists, the initial idx list is `(N, max_neighbors_per_cell)` before pruning. In this case, the pruning function is the same but it costs O(N * max_neighbors_per_cell) time which can be much faster for large systems. Typically the max neighbors per cell is a constant independent of N.

**Comment by oliverdutton** [comment]
Yes. All makes sense. (Though `max_neighbors_per_cell` being ~24x the number of particles in the system seems a bit excessive.)

Tests added, you'll see a simple base wrapping of the function `acceptable_id_pair` to work with the neighbor lists.

Base function:
```python
def acceptable_id_pair(id1, id2):
  # Don't allow particles to have an interaction when their id's are closer than 3 (eg disabling 1-2 and 1-3 interactions)
  return jnp.abs(id1-id2)>3
```

Wrapper:
```python
def mask_id_based(idx: Array, ids: Array, mask_val: int, _acceptable_id_pair: Callable) -> Array:
  '''
  _acceptable_id_pair mapped to act upon the neighbor list where:
      - index of particle 1 is in index in the first dimension of the array
      - index of particle 2 is given by the value in the array
  '''
  @partial(vmap, in_axes=(0,0,None))
  def acceptable_id_pair(idx, id1, ids):
    id2 = ids.at[idx].get()
    return vmap(_acceptable_id_pair, in_axes=(None,0))(id1,id2)
  mask = acceptable_id_pair(idx, ids, ids) # Note the id's twice, one iterated, one is passed fully in (to then be coverted to the second, value-based, index)
  return jnp.where(
    mask,
    idx,
    mask_val
  )

ids = jnp.arange(n_particles) # id is just particle index here.
mask_val = n_particles
custom_mask_function = partial(mask_id_based, ids=ids, mask_val=mask_val, _acceptable_id_pair=acceptable_id_pair)


```

**Comment by oliverdutton** [comment]
So @sschoenholz is this now ready for merging?

**Review by sschoenholz — COMMENTED** [review]
Sorry for the long delay, we took a few weeks off for winter break! This change is looking great and as soon as you fix one minor formatting issue in the tests, I'm happy to merge.

File: jax_md/partition.py [hunk 1]
```diff
@@ -544,6 +544,7 @@ def neighbor_list(displacement_or_metric: DisplacementOrMetricFn,
                   capacity_multiplier: float = 1.25,
                   disable_cell_list: bool = False,
                   mask_self: bool = True,
+                  custom_mask_function: Optional[Callable[[Array], Array]] = None,
```
Comment (oliverdutton) [inline jax_md/partition.py:547]: This line breaks the 80 char limit, is this okay?

File: tests/partition_test.py [hunk 2]
```diff
@@ -363,5 +363,52 @@ def test_cell_list_overflow(self):
     self.assertTrue(neighbors.did_buffer_overflow)
     self.assertEqual(neighbors.idx.dtype, jnp.int32)
 
+def test_custom_mask_function(self):
```
[the remark this answers is not in the document]
Comment (oliverdutton) [reply tests/partition_test.py]: Yes, it's not being run. Fixed in new commits.

File: tests/partition_test.py [hunk 3]
```diff
@@ -363,5 +363,52 @@ def test_cell_list_overflow(self):
...
+    mask = acceptable_id_pair(idx, ids, ids)
+    return jnp.where(
+      mask,
+      idx,
+      mask_val
+    )
+
+  ids = jnp.arange(n_particles) # id is just particle index here.
+  mask_val = n_particles
+  custom_mask_function = partial(mask_id_based, ids=ids, mask_val=mask_val, _acceptable_id_pair=acceptable_id_pair)
```
[the remark this answers is not in the document]
Comment (oliverdutton) [reply tests/partition_test.py]: Of course, fixed in new commits.

File: jax_md/partition.py (hunk 1 above)
Comment (sschoenholz) [reply jax_md/partition.py:547]: Hm, maybe we can define an alias `MaskFn = Callable[[Array], Array]`? (After this change I'm happy to merge, thanks for your patience!)

Comment (oliverdutton) [reply jax_md/partition.py:547]: Sounds good, I'll push that. 

It will then probably fail in linting, I need to make it conform before it'll be ready ready.

**Comment by oliverdutton** [comment]
Okay, now it should be in a shape to merge.
